src/Reconcile.js

```diff
@@ -0,0 +1,189 @@
+import * as Constant from './Constant.js'
+import * as Utility from './Utility.js'
+import * as Commit from './Commit.js'
+import * as Component from './Component.js'
+
+/**
+ * @param {object} host
+ * @param {object} parent
+ * @param {object} element
+ * @param {object} snapshot
+ * @param {number} index
+ */
+export function update (host, parent, element, snapshot, index) {
+	if (element.type !== snapshot.type) {
+		if (element.uuid !== Constant.thenable || snapshot.uuid !== Constant.thenable) {
+			return Commit.replace(host, parent, element, snapshot, index)
+		}
+	}
+
+	switch (element.uuid) {
+		case Constant.text:
+		case Constant.empty:
+		case Constant.comment:
+			return content(element, element.children, snapshot.children)
+		case Constant.component:
+			return Component.update(host, parent, element, snapshot, Constant.object, Constant.update)
+		case Constant.thenable:
+			return Commit.thenable(host, parent, element, index, element.type = snapshot.type)
+		default:
+			children(host, element, element.children, snapshot.children)
+			props(element, element.props, snapshot.props)
+	}
+}
+
+/**
+ * @param {object} element
+ * @param {(string|number)} a
+ * @param {(string|number)} b
+ */
+export function content (element, a, b) {
+	if (a != b) {
+		Commit.content(element, element.children = b)
+	}
+}
+
+/**
+ * @param {object} element
+ * @param {object} a
+ * @param {object} b
+ */
+export function props (element, a, b) {
+	if (a !== b) {
+		Commit.props(element, object(a, element.props = b), element.xmlns, Constant.update)
+	}
+}
+
+/**
+ * @param {object} a
+ * @param {object} b
+ * @return {object?}
+ */
+export function object (a, b) {
+	var size = 0
+	var diff = {}
+
+	for (var key in a) {
+		if (!Utility.has(b, key)) {
+			diff[++size, key] = null
+		}
+	}
+
+	for (var key in b) {
+		var prev = a[key]
+		var next = b[key]
+
+		if (prev !== next) {
+			if (typeof next !== 'object' || next === null) {
+				diff[++size, key] = next
+			} else if (next = object(prev || {}, next)) {
+				diff[++size, key] = next
+			}
+		}
+	}
+
+	if (size) {
+		return diff
+	}
+}
+
+/**
+ * @param {object} host
+ * @param {object} parent
+ * @param {object} a
+ * @param {number} b
+ */
+export function children (host, parent, a, b) {
+	var prev = parent
+	var next = parent
+	var alen = a.length
+	var blen = b.length
+	var aend = alen - 1
+	var bend = blen - 1
+	var min = bend > aend ? aend : bend
+	var max = bend > aend ? bend : aend
+	var idx = 0
+	var ptr = 0
+
+	// step 1, prefix/suffix
+	outer: {
+		while ((prev = a[idx]).key === (next = b[idx]).key) {
+			update(host, parent, prev, next, idx)
+			if (++idx > min) { break outer }
+		}
+		while ((prev = a[aend]).key === (next = b[bend]).key) {
+			update(host, parent, a[max] = prev, next, bend)
```

Comment:
We want to avoid splicing when unnecessary so depending on which side is bigger we move(in-place) the values by the time we reach un/mount phase(assuming we do) there is an assumed section between two points that has dummy values or holes if the statement holds true. i.e

```
[2, 3, 4]
[0, 2, 3, 4]
```
The zero invalidates the prefix loop, the suffix loop has 3 matching keys, but space for it(zero) is not available, the first array this transitions to

```
[2, 2, 3, 4]
[0, 2, 3, 4]
```

Where the first 2 is the assumed dummy value.